main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala

```diff
@@ -100,13 +100,14 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
     // 3. Create Enum Type info
     val allEnums : List[(Type, (String, Type))] = root.definitions.values.flatMap(x => CodegenHelper.findEnumCases(x.exp)).toList
 
-    val enumTypeInfo: Map[(Type, String), QualName] = allEnums.map{ case (tpe, (name, subType)) =>
+    val enumTypeInfo: Map[(Type, String), (QualName, Boolean)] = allEnums.map{ case (tpe, (name, subType)) =>
```

Comment:
I am wondering if passing the `ExecutableAst.Case` ast node would not simplify things?

Then you could have a helper method `def isSingleton(caze: ExecutableAst.Case)` basically trying to reuse the AST ast much as possible. I understand that for the specialized types you need to carry more information than the enum/case, but having these nodes available might still simplify things.